File: PeakTrailPlatform/server/run-code.mjs
// Deterministic 4-character run code, shared by the recorder (C# RunCode.cs) and
// this server. Every modded client in the same PEAK run derives the SAME code from
// the room-shared RunManager.RunId, so the server can merge all their uploads into
// one run without any client-side coordination.
import { createHash } from "node:crypto";

export const CODE_ALPHABET = "23456789abcdefghjkmnpqrstuvwxyz"; // no 0/o/1/i/l — 31 chars
export const CODE_LENGTH = 4;
export const CODE_SPACE = CODE_ALPHABET.length ** CODE_LENGTH; // 923521
export const MAX_DERIVATION_ATTEMPTS = 8;

const GUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

/** .NET Guid "D" form: lowercase, hyphenated; anything else is trimmed lowercase. */
export function normalizeRunId(runId) {
  const text = String(runId ?? "").trim().toLowerCase().replace(/^[{[]|[}\]]$/g, "");
  if (GUID_PATTERN.test(text)) return text;
  const compact = text.replaceAll("-", "");
  if (/^[0-9a-f]{32}$/.test(compact)) {
    return `${compact.slice(0, 8)}-${compact.slice(8, 12)}-${compact.slice(12, 16)}-${compact.slice(16, 20)}-${compact.slice(20)}`;
  }
  return text;
}

/** First four code characters for this run; `attempt` walks a deterministic
 * fallback sequence when the server reports a code collision (409). */
export function deriveRunCode(runId, attempt = 0) {
  const normalized = normalizeRunId(runId);
  if (!normalized) return "";
  const digest = createHash("sha256").update(`${normalized}:${attempt}`, "utf8").digest();
  // Explicit little-endian u32 so the C# mirror never depends on machine endianness.
  // >>> 0 reinterprets the signed int32 bitwise result as unsigned, matching C# uint.
  const value = (digest[0] | (digest[1] << 8) | (digest[2] << 16) | (digest[3] << 24)) >>> 0;
  let rest = value % CODE_SPACE;
  const chars = new Array(CODE_LENGTH);
  for (let index = CODE_LENGTH - 1; index >= 0; index -= 1) {
    chars[index] = CODE_ALPHABET[rest % CODE_ALPHABET.length];
    rest = Math.floor(rest / CODE_ALPHABET.length);
  }
  return chars.join("");
}

export function isValidRunCode(code) {
  return typeof code === "string" && code.length === CODE_LENGTH
    && [...code].every((character) => CODE_ALPHABET.includes(character));
}

/** The server "confirms" a uuid by re-deriving it from the runId. */
export function runCodeMatches(runId, code) {
  for (let attempt = 0; attempt < MAX_DERIVATION_ATTEMPTS; attempt += 1) {
    if (deriveRunCode(runId, attempt) === code) return true;
  }
  return false;
}

File: PeakTrailRecorder/src/PeakTrailRecorder/RunCode.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace PeakTrailRecorder;

/// <summary>
/// Deterministic 4-character run code, byte-for-byte identical to the server's
/// reference implementation (PeakTrailPlatform/server/run-code.mjs). Every modded
/// client in the same PEAK run derives the SAME code from the room-shared
/// RunManager.RunId, so the relay can merge all their uploads into one run
/// without any client-side coordination.
/// </summary>
internal static class RunCode
{
    public const string Alphabet = "23456789abcdefghjkmnpqrstuvwxyz"; // no 0/o/1/i/l — 31 chars
    public const int Length = 4;
    public const int MaxDerivationAttempts = 8;
    private static readonly uint CodeSpace = (uint)Math.Pow(Alphabet.Length, Length); // 923521

    public static string Normalize(string? runId)
    {
        string text = (runId ?? string.Empty).Trim().ToLowerInvariant();
        if (text.Length >= 2 && ((text[0] == '{' && text[^1] == '}') || (text[0] == '[' && text[^1] == ']')))
        {
            text = text[1..^1];
        }
        if (Guid.TryParse(text, out Guid guid)) return guid.ToString("D");
        string compact = text.Replace("-", string.Empty, StringComparison.Ordinal);
        if (compact.Length == 32 && IsHex(compact))
        {
            return $"{compact[..8]}-{compact.Substring(8, 4)}-{compact.Substring(12, 4)}-{compact.Substring(16, 4)}-{compact.Substring(20)}";
        }
        return text;
    }

    /// <summary>First code for this run; <paramref name="attempt"/> walks a
    /// deterministic fallback sequence when the server reports a code collision.</summary>
    public static string Derive(string? runId, int attempt = 0)
    {
        string normalized = Normalize(runId);
        if (normalized.Length == 0) return string.Empty;
        byte[] digest;
        // SHA256.HashData (.NET 6+) is unavailable on netstandard2.1, which this
        // file must also compile against for the game plugin.
#pragma warning disable CA1850
        using (SHA256 sha = SHA256.Create())
        {
            digest = sha.ComputeHash(Encoding.UTF8.GetBytes($"{normalized}:{attempt}"));
        }
#pragma warning restore CA1850
        // Explicit little-endian u32; must match the JS `>>> 0` reinterpretation.
        uint value = digest[0] | (uint)digest[1] << 8 | (uint)digest[2] << 16 | (uint)digest[3] << 24;
        uint rest = value % CodeSpace;
        Span<char> chars = stackalloc char[Length];
        for (int index = Length - 1; index >= 0; index--)
        {
            chars[index] = Alphabet[(int)(rest % (uint)Alphabet.Length)];
            rest /= (uint)Alphabet.Length;
        }
        return new string(chars);
    }

    /// <summary>The server "confirms" a code by re-deriving it from the runId.</summary>
    public static bool Matches(string? runId, string? code)
    {
        if (string.IsNullOrEmpty(code) || code.Length != Length) return false;
        for (int attempt = 0; attempt < MaxDerivationAttempts; attempt++)
        {
            if (string.Equals(Derive(runId, attempt), code, StringComparison.Ordinal)) return true;
        }
        return false;
    }

    private static bool IsHex(string text)
    {
        foreach (char character in text)
        {
            bool hex = character is >= '0' and <= '9' or >= 'a' and <= 'f';
            if (!hex) return false;
        }
        return true;
    }
}
